### ISSUE TITLE
Any plans to support rich text/cards response validation

### ISSUE THREAD

**Issue by Endeavour-BRM** [codeforequity-at/botium-core#58]
Microsoft Bot framework has multiple rich cards supported as response, do we have any options to validate the response content of rich cards.

And also is there any ways to validate the response from bot that is basically a button list?

**Comment by codeforequity-at** [comment]
Did you succeed in this or can I help ?

**Comment by Endeavour-BRM** [comment]
@codeforequity I have similar use case I have an adaptive card. Bot response looks like this
{
  "type": "message",
  "id": "Gp4RyH0BbX5JTTUc9feTCS|0000015",
  "timestamp": "2018-07-04T11:16:42.716263Z",
  "localTimestamp": "2018-07-04T11:16:42.603+00:00",
  "channelId": "directline",
  "from": {
   ....
  },
  "conversation": {
    "id": "Gp4RyH0BbX5JTTUc9feTCS"
  },
  "inputHint": "ignoringInput",
  "attachments": [
    {
     ......
  ],
  "replyToId": "Gp4RyH0BbX5JTTUc9feTCS|0000014"
}

The ID, replyToId and conversation id value changes for each session. Based on article I figured out that I can add regex capability in Botium settings "SCRIPTING_MATCHING_MODE": "regexp". Can you please help me with the regexp syntax that I have to add it in convo file in this json section so that the ID in the above json is considered as regex or ignored during assertion.

Thanks.

**Comment by codeforequity-at** [comment]
Botium automatically detects JSON in the convo files and compares against the whole message structure from the source system. So instead of adding a simple text there, copy&paste the whole structure from above.
Botium does a partial compare only: you can remove everything you don't want to be validated, the sessionId attribute, the replyToId attribute, the id attributes, the timestamp attributes, ... and only validate the attachments node.

**Comment by Endeavour-BRM** [comment]
@codeforequity-at Thanks that worked.

**Comment by codeforequity-at** [comment]
Great news!

**Comment by cr33dx** [comment]
> Basically, this is supported right now: it is possible to assert the native bot response against JSON structures by just adding some JSON code to the script files (see for example this one: https://github.com/codeforequity-at/testmybot/blob/master/samples/facebook/spec/convo/should_answer_to_postback.convo.txt)
> 
> What's native of course depends on the underlying bot technology you are using, so the best approach would be to enable debug logging ("export DEBUG=botium*"), so you see the native bot response in the console window when running a Botium script, and you can then adapt your assertions step by step.
> 
> To be honest, currently this is rather cumbersome, and this will get better in future, by adding some assertion abstraction for rich messages in the most important platforms. That's the reason why this isn't really documented.

The said github link is 404 . is there a way to do this now?

**Comment by codeforequity-at** [comment]
yes, see https://medium.com/@floriantreml/botium-in-a-nutshell-part-2-writing-chatbot-tests-8726204c8da6

**Comment by cr33dx** [comment]
Can I fill up a form?
And I'm getting this response
![qq](https://user-images.githubusercontent.com/8844755/54333520-be8cd100-4648-11e9-9be2-a5890b21a569.png)

Whereas I want a response like this where I can at least fill up the form

![Screenshot (10)](https://user-images.githubusercontent.com/8844755/54333578-0ca1d480-4649-11e9-9eaa-432727f2cb69.png)

**Comment by codeforequity-at** [comment]
Interesting. Right now we only support button clicks. I added it to our backlog to support other form elements as well. You can expect a solution in one of the next builds.

**Comment by cr33dx** [comment]
yay, so I just need to re-run docker-compose up for botium box once you commit changes?

**Comment by codeforequity-at** [comment]
yes. To be exact, it will be a "docker-compose pull && docker-compose up".

**Comment by vijaysimh** [comment]
Hello am facing the problem in Adaptive cards as seen in this discussion ,me to getting the same response as "Do you want   to delete order permanently from XXX? or Do you want to close   XXX?,Delete Order,Close Order" -BOT_SENDS_CARDS (emulator) as response

As per the cmd:botium-cli run -Its capturing response as 'Do you want   to delete order permanently from XXX? or Do you want to close   XXX?'
Why ,Delete Order,Close Order Options are missing from response or i need to make changes in json file.
Delete Order and Close Order are comma separated and am currently using capabilities as below:
Water Fall dialog:
{
  "botium": {
    "Capabilities": {
      "PROJECTNAME": "BotTesting",
      "CONTAINERMODE": "directline3",
	  "CLEANUPTEMPDIR":true,  
      "DIRECTLINE3_SECRET": "XXXX",
	  "DIRECTLINE3_WEBSOCKET": true,
      "DIRECTLINE3_POLLINGINTERVAL": 1000,
	  "DIRECTLINE3_BUTTON_TYPE": "message",
      "DIRECTLINE3_BUTTON_VALUE_FIELD": "text",
      "DIRECTLINE3_HANDLE_ACTIVITY_TYPES": [ "message", "event" ],
	  "WAITFORBOTTIMEOUT":10000,
	  	  "SCRIPTING_FORMAT": "xlsx",
		  "SCRIPTING_XLSX_STARTROW": 2,
		  "SCRIPTING_XLSX_STARTCOL": 1,
	 "SCRIPTING_XLSX_SHEETNAMES":"HelloUtt",
	 "SCRIPTING_XLSX_SHEETNAMES_UTTERANCES":"Utterances",
	 "SCRIPTING_ENABLE_MEMORY": true
    },
      "Sources": {},
      "Envs": {
        "NODE_TLS_REJECT_UNAUTHORIZED": 0
      }
  }
}

PLEASE SUGGEST

**Comment by codeforequity-at** [comment]
@vijaysimh please open a separate ticket, thanks!
